sdk/core/azure-core/src/main/java/com/azure/core/implementation/AccessTokenCache.java

```diff
@@ -161,6 +170,83 @@ private Supplier<Mono<? extends AccessToken>> retrieveToken(TokenRequestContext
+
+            // Check if the incoming token request context is different from the cached one. A different
+            // token request context, requires to fetch a new token as the cached one won't work for the
+            // passed in token request context.
+            boolean forceRefresh = (checkToForceFetchToken && checkIfForceRefreshRequired(tokenRequestContext))
+                || this.tokenRequestContext == null;
+
+            Supplier<AccessToken> tokenSupplier = () ->
+                tokenCredential.getTokenSync(this.tokenRequestContext);
+
+            if (forceRefresh) {
+                this.tokenRequestContext = tokenRequestContext;
+                tokenRefresh = tokenSupplier;
+                fallback = null;
+            } else if (cachedToken != null && !shouldRefresh.test(cachedToken)) {
+                // fresh cache & no need to refresh
+                tokenRefresh = null;
+                fallback = cachedToken;
+            } else if (cachedToken == null || cachedToken.isExpired()) {
+                // no token to use
+                // refresh immediately
+                tokenRefresh = tokenSupplier;
+
+                // cache doesn't exist or expired, no fallback
+                fallback = null;
+            } else {
+                // token available, but close to expiry
+                if (now.isAfter(cache.getNextTokenRefresh())) {
+                    // refresh immediately
+                    tokenRefresh = tokenSupplier;
+                } else {
+                    // still in timeout, do not refresh
+                    tokenRefresh = null;
+                }
+                // cache hasn't expired, ignore refresh error this time
+                fallback = cachedToken;
+            }
+
+            try {
+                if (tokenRefresh != null) {
+                    AccessToken token = tokenRefresh.get();
+                    buildTokenRefreshLog(LogLevel.INFORMATIONAL, cachedToken, now)
+                        .log("Acquired a new access token.");
+                    OffsetDateTime nextTokenRefreshTime = OffsetDateTime.now().plus(REFRESH_DELAY);
+                    AccessTokenCacheInfo updatedInfo = new AccessTokenCacheInfo(token, nextTokenRefreshTime);
+                    this.cacheInfo.set(updatedInfo);
+                    return token;
+                } else {
+                    return fallback;
+                }
+            } catch (Throwable error) {
+                buildTokenRefreshLog(LogLevel.ERROR, cachedToken, now)
+                    .log("Failed to acquire a new access token.", error);
+                OffsetDateTime nextTokenRefreshTime = OffsetDateTime.now();
+                AccessTokenCacheInfo updatedInfo = new AccessTokenCacheInfo(cachedToken, nextTokenRefreshTime);
+                this.cacheInfo.set(updatedInfo);
+                if (fallback != null) {
+                    return fallback;
+                }
+                throw error;
```

Comment:
Should this be thrown through the `ClientLogger`?

Reply:
we do log it above I believe